src/main/java/de/zalando/aruha/nakadi/service/CursorsCommitService.java

```diff
@@ -0,0 +1,120 @@
+
+    private static final Charset CHARSET = Charset.forName("UTF-8");
+
+    private final ZooKeeperHolder zkHolder;
+    private final TopicRepository topicRepository;
+    private final SubscriptionDbRepository subscriptionRepository;
+    private final ZooKeeperLockFactory zkLockFactory;
+    private final ZkSubscriptionClientFactory zkSubscriptionClientFactory;
+    private final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory;
+
+    public CursorsCommitService(final ZooKeeperHolder zkHolder,
+                                final TopicRepository topicRepository,
+                                final SubscriptionDbRepository subscriptionRepository,
+                                final ZooKeeperLockFactory zkLockFactory,
+                                final ZkSubscriptionClientFactory zkSubscriptionClientFactory,
+                                final SubscriptionKafkaClientFactory subscriptionKafkaClientFactory) {
+        this.zkHolder = zkHolder;
+        this.topicRepository = topicRepository;
+        this.subscriptionRepository = subscriptionRepository;
+        this.zkLockFactory = zkLockFactory;
+        this.zkSubscriptionClientFactory = zkSubscriptionClientFactory;
+        this.subscriptionKafkaClientFactory = subscriptionKafkaClientFactory;
+    }
+
+    public boolean commitCursors(final String subscriptionId, final List<Cursor> cursors)
+            throws InvalidCursorException, NoSuchSubscriptionException, ServiceUnavailableException {
+
+        final Subscription subscription = subscriptionRepository.getSubscription(subscriptionId);
+        final String eventType = subscription.getEventTypes().iterator().next();
+
+        createSubscriptionInZkIfNeeded(subscription);
+        topicRepository.validateCommitCursors(eventType, cursors);
+
+        boolean allCommitted = true;
+        for (final Cursor cursor : cursors) {
+            allCommitted = allCommitted && commitCursor(subscriptionId, eventType, cursor);
+        }
+        return allCommitted;
+    }
+
+    private boolean commitCursor(final String subscriptionId, final String eventType, final Cursor cursor)
+            throws ServiceUnavailableException, NoSuchSubscriptionException, InvalidCursorException {
+
+        final String offsetPath = format("/nakadi/subscriptions/{0}/topics/{1}/{2}/offset",
+                subscriptionId, eventType, cursor.getPartition());
+        try {
+            return runLocked(() -> {
+                final String currentOffset = new String(zkHolder.get().getData().forPath(offsetPath), CHARSET);
+                if (topicRepository.compareOffsets(cursor.getOffset(), currentOffset) > 0) {
+                    zkHolder.get().setData().forPath(offsetPath, cursor.getOffset().getBytes(CHARSET));
+                    return true;
+                } else {
+                    return false;
+                }
+            }, zkLockFactory.createLock(offsetPath));
+        } catch (final IllegalArgumentException e) {
```

Comment:
I also thought of that. The only conditions when it can happen - is if the code is changed. And if the code is changed - then probably 422 will be more appropriate than 500